pkg/geo/geogen/geogen.go

```diff
@@ -103,75 +176,86 @@ func RandomLineString(
 			minTrunc, maxTrunc = maxTrunc, minTrunc
 		}
 	}
-	return geom.NewLineString(geom.XY).MustSetCoords(randCoords[minTrunc:maxTrunc]).SetSRID(int(srid))
+	return geom.NewLineString(layout).MustSetCoords(randCoords[minTrunc:maxTrunc]).SetSRID(int(srid))
 }
 
 // RandomPolygon generates a random Polygon.
 func RandomPolygon(
-	rng *rand.Rand, minX float64, maxX float64, minY float64, maxY float64, srid geopb.SRID,
+	rng *rand.Rand, randomBounds RandomGeomBounds, srid geopb.SRID, layout geom.Layout,
 ) *geom.Polygon {
+	layout = RandomLayout(rng, layout)
+	// 10% chance to generate an empty polygon.
+	if rng.Intn(10) == 0 {
+		return geom.NewPolygon(layout).SetSRID(int(srid))
+	}
 	// TODO(otan): generate random holes inside the Polygon.
 	// Ideas:
 	// * We can do something like use 4 arbitrary points in the LinearRing to generate a BoundingBox,
 	//   and re-use "PointInLinearRing" to generate N random points inside the 4 points to form
 	//   a "sub" linear ring inside.
 	// * Generate a random set of polygons, see which ones they fully cover and use that.
-	return geom.NewPolygon(geom.XY).MustSetCoords([][]geom.Coord{
-		RandomValidLinearRingCoords(rng, 3+rng.Intn(10), minX, maxX, minY, maxY),
+	return geom.NewPolygon(layout).MustSetCoords([][]geom.Coord{
+		RandomValidLinearRingCoords(rng, 3+rng.Intn(10), randomBounds, layout),
 	}).SetSRID(int(srid))
 }
 
-// RandomGeomT generates a random geom.T object within the given bounds and SRID.
+// RandomGeomT generates a random geom.T object with the given layout within the
+// given bounds and SRID.
 func RandomGeomT(
-	rng *rand.Rand, minX float64, maxX float64, minY float64, maxY float64, srid geopb.SRID,
+	rng *rand.Rand, randomBounds RandomGeomBounds, srid geopb.SRID, layout geom.Layout,
 ) geom.T {
-	shapeType := validShapeTypes[rng.Intn(len(validShapeTypes))]
+	layout = RandomLayout(rng, layout)
+	shapeType := RandomShapeType(rng)
 	switch shapeType {
 	case geopb.ShapeType_Point:
-		return RandomPoint(rng, minX, maxX, minY, maxY, srid)
+		return RandomPoint(rng, randomBounds, srid, layout)
 	case geopb.ShapeType_LineString:
-		return RandomLineString(rng, minX, maxX, minY, maxY, srid)
+		return RandomLineString(rng, randomBounds, srid, layout)
 	case geopb.ShapeType_Polygon:
-		return RandomPolygon(rng, minX, maxX, minY, maxY, srid)
+		return RandomPolygon(rng, randomBounds, srid, layout)
 	case geopb.ShapeType_MultiPoint:
-		// TODO(otan): add empty points.
-		ret := geom.NewMultiPoint(geom.XY).SetSRID(int(srid))
-		num := 1 + rng.Intn(10)
+		ret := geom.NewMultiPoint(layout).SetSRID(int(srid))
+		// 10% chance to generate an empty multipoint.
```

Comment:
does this? it feels more like "each point has a chance to generate a random point inside".
do you need another check here for this and below that it will generate a random multi point?
also if we end up doing that, can you split all these out into individual functions?

Reply:
I think it should since I changed it so `num` will be in the range [0, 10) instead of [1, 11). If it's 0 it should cause no points to be pushed (i.e. an empty multipoint).

Reply:
d'oh, i see

Reply:
(if you could separate them out into separate functions that'd be great, maybe a comment on how it's 10% would be good)